MetaMask/metamask-mobile
# feat: segment data deletion (4/4)

## **Description**

1. What is the reason for the change?
User has to be able to request deletion of their metrics data stored in new segment system.

2. What is the improvement/solution?

- adds `createDataDeletionTask` to `MetaMetrics` to call the deletion API in MetaMetrics class for method [create source regulation](https://docs.segmentapis.com/tag/Deletion-and-Suppression#operation/createSourceRegulation)
- adds the `checkDataDeletionTaskStatus` to `MetaMetrics` to call deletion API [get regulation method](https://docs.segmentapis.com/tag/Deletion-and-Suppression#operation/getRegulation) 
- calls the `createDataDeletionTask` and `checkDataDeletionTaskStatus` from `DeleteMetaMetricsData.tsx` when user touches the delete button.
- extract the settings section logic to display allow deletion to a well tested hook.
- Remove `Analytics.js` implementation of these features and point only to the new `MetaMetrics` ones.
- update unit tests
- rename in `storage.js`:
  - `METAMETRICS_SEGMENT_REGULATION_ID` -> `METAMETRICS_DELETION_REGULATION_ID`

## **Related issues**

Fixes https://github.com/MetaMask/mobile-planning/issues/1403

## **Manual testing steps**

> [!NOTE]
> Testing is non regression testing as the feature should work exactly the same as before.

There's 3 cases here:
- initial deletion
- user opted IN and goes back on settings once a deletion request has been initiated
- user opted OUT and goes back on settings once a deletion request has been initiated


```gherkin
Feature: opted-in or out user deletes MetaMetrics Data for the first time
  Given the app is installed and setup
  And user goes into settings / security & privacy
  And no deletion request has been done before
  And "Delete Metametrics data" section text is "This will delete historical 
    MetaMetrics data associated with your wallet.[...] View the ConsenSys Privacy Policy."

  When user touches the "delete metametrics" button
  And approves deletion

  Then "Delete Metametrics data" section text changes to "You initiated this action on [current date]. 
    This process can take up to 60 days. View the ConsenSys Privacy Policy."
  And "delete metametrics" button is now disabled
  And user still continues to be tracked or not depending on opt-out statusbut
  And the tracking strategy is not changed by the deletion action

Feature: opted-in user is back on MetaMetrics settings screen after deletion
  Given user has opted-in for metrics
  And user asked for metrics data deletion
  And user exit settings screen

  When user comes back to settings / security & privacy
  And users scrolls down to the delete metametrics section

  Then "Delete Metametrics data" section text is "This will delete historical 
    MetaMetrics data associated with your wallet.[...] View the ConsenSys Privacy Policy."
  And user can click the "delete metametrics" button again

Feature: opted-out user is back on MetaMetrics settings screen after deletion
  Given user has opted-out for metrics
  And user asked for metrics data deletion
  And user exit settings screen

  When user comes back to settings / security & privacy
  And users scrolls down to the delete metametrics section

  Then "Delete Metametrics data" section text is "You initiated this action on [current date]. 
    This process can take up to 60 days. View the ConsenSys Privacy Policy."
  And "delete metametrics" button is disabled
  And user can NOT click the "delete metametrics" button again until data deletion task is finished
```

## **Screenshots/Recordings**

https://github.com/MetaMask/metamask-mobile/assets/4677568/8fb3201c-d07e-4903-96f9-c3217f44f9e4

"Security and Privacy" settings "Delete Metametrics data" section BEFORE deletion request:
<img width="304" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/6f2bd6e2-0fd1-4cf2-b706-c288a1abe818">

"Security and Privacy" settings "Delete Metametrics data" section AFTER deletion request:
<img width="304" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/30ba3c72-1e4b-4749-9da1-98364cd6f92f">

"Delete MetaMetrics Data Request Submitted" event tracked on Segment and visible in the dashboard:
```json
{
  "context": {
    "library": {
      "name": "source-functions",
      "version": "1.0.0"
    },
    "location": {
      "api_processing_time_ms": 520,
      "country_code": "FR",
      "region": "NOR",
      "timezone": "Europe/Paris"
    }
  },
  "event": "Delete MetaMetrics Data Request Submitted",
  "integrations": {},
  "messageId": "e09b2aac-dd78-44d5-8c48-9ee1464836b1",
  "originalTimestamp": "2023-12-13T23:50:58.790Z",
  "properties": {
    "device_model": "Apple iPhone14,2",
    "os": "ios",
    "os_version": "17.0.1"
  },
  "receivedAt": "2023-12-13T23:51:31.609Z",
  "sentAt": "2023-12-13T23:51:31.609Z",
  "timestamp": "2023-12-13T23:50:58.790Z",
  "type": "track",
  "userId": "35a6b109-d2b4-4d1f-a3ee-7dd454bd8667",
  "writeKey": "REDACTED"
}
```

Deletion Request appears as "INITIALIZED" in the Segment Deletion Requests dashboard.
You can make sure this is your request by matching the ID with the ID that should have been sent to segment during the onboarding and by the date and time of the request.
<img width="1381" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/b3112893-a1b1-42ba-97e8-9f075f3c8044">


### **Before**

No change, this PR only makes this work with Segment, but UX should be the same

### **After**

No change, this PR only makes this work with Segment, but UX should be the same

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
